Testing infra and some fixes

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #162166
* #162089
* #162013
* __->__ #161977
* #161653

This PR is quite large in that it covers most of rough edges in the new strict export flow:
1. Handle nn_module_stack correctly now that we are tracing wrapper module 
2. module_call_spec needs to get queried from source directly because we are not running the bytecode anymore. 
3. Correct input and output handling. 



cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

Differential Revision: [D81589677](https://our.internmc.facebook.com/intern/diff/D81589677)

tugsbayasgalan committed:
Testing infra and some fixes

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Testing infra and some fixes"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Testing infra and some fixes"


This PR is quite large in that it covers most of rough edges in the new strict export flow:
1. Handle nn_module_stack correctly now that we are tracing wrapper module 
2. module_call_spec needs to get queried from source directly because we are not running the bytecode anymore. 



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan commented on torch/_dynamo/functional_export.py:
```diff
@@ -67,72 +312,133 @@ def inner(*args: Any, **kwargs: Any):
+
+        constraint_violation_error = None
+        try:
+            check_fn = out.dynamo_output.build_guards(
```
cc: @avikchaudhuri  I am just using guards here instead of calling non_strict equivalent. They look quite similar to each other actually.

tugsbayasgalan replied:
@avikchaudhuri I feel the new data structure should be able to generate both guards and bytecodes. Today, aot_precompile uses both bytecode (that is generated from output_graph directly) and output_graph_guard_state for guards. In my mental model, both of these should come from single data structure instead of us emitting the end artifacts seperately. (Today's output_graph is the closest we have to this and we need to trim it down). cc: @anijain2305

tugsbayasgalan committed:
Update on "Testing infra and some fixes"


This PR is quite large in that it covers most of rough edges in the new strict export flow:
1. Handle nn_module_stack correctly now that we are tracing wrapper module 
2. module_call_spec needs to get queried from source directly because we are not running the bytecode anymore. 
3. Correct input and output handling. 



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Testing infra and some fixes"


This PR is quite large in that it covers most of rough edges in the new strict export flow:
1. Handle nn_module_stack correctly now that we are tracing wrapper module 
2. module_call_spec needs to get queried from source directly because we are not running the bytecode anymore. 
3. Correct input and output handling. 



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zhxchen17 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -900,7 +901,7 @@ class CaptureOutput:
 
     dynamo_output: DynamoOutput
-    backend_input: BackendInput
+    backend_input: Optional[BackendInput]
```
```suggestion
    backend_input: BackendInput
```

zhxchen17 replied:
@tugsbayasgalan add a comment to indicate "None" means dynamo doesn't capture any graph.

avikchaudhuri replied:
In torch.compile, how is this handled? Do we pass an empty graph to the backend?

zhxchen17 replied:
we just never ever call a backend in that case and return only the bytecodes

zhxchen17 commented on torch/_dynamo/convert_frame.py:
```diff
@@ -913,7 +914,9 @@ class FrameInfo:
 def fullgraph_capture(
-    frame: FrameInfo, _is_export_deprecated_do_not_use: bool = False
+    frame: FrameInfo,
+    _is_export_deprecated_do_not_use: bool = False,
+    _export_constraints: Optional[list[Constraint]] = None,
```
```suggestion
    constraints: Optional[list[Constraint]] = None,
    *,
    _is_export_deprecated_do_not_use: bool = False,
```
constraints doesn't have to be export only

tugsbayasgalan committed:
Update on "Testing infra and some fixes"


This PR is quite large in that it covers most of rough edges in the new strict export flow:
1. Handle nn_module_stack correctly now that we are tracing wrapper module 
2. module_call_spec needs to get queried from source directly because we are not running the bytecode anymore. 
3. Correct input and output handling. 



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D81589677](https://our.internmc.facebook.com/intern/diff/D81589677)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Testing infra and some fixes"


This PR is quite large in that it covers most of rough edges in the new strict export flow:
1. Handle nn_module_stack correctly now that we are tracing wrapper module 
2. module_call_spec needs to get queried from source directly because we are not running the bytecode anymore. 
3. Correct input and output handling. 



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D81589677](https://our.internmc.facebook.com/intern/diff/D81589677)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Testing infra and some fixes"


This PR is quite large in that it covers most of rough edges in the new strict export flow:
1. Handle nn_module_stack correctly now that we are tracing wrapper module 
2. module_call_spec needs to get queried from source directly because we are not running the bytecode anymore. 
3. Correct input and output handling. 



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D81589677](https://our.internmc.facebook.com/intern/diff/D81589677)

[ghstack-poisoned]
